Add constant to FormulaEngine arithmetic

matthias-wende-frequenz opened the pull request:
We can now use constants for all formula engine operations.

**Example:**
```python
scaled_grid_power = (microgrid.logical_meter().grid_power * 5).build()
```

matthias-wende-frequenz committed:
Add token type CONSTANT

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Support constants in arithmetic FormulaEngine operations

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

shsms commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -715,6 +721,10 @@ def _push(
         # pylint: disable=protected-access
         if isinstance(other, (FormulaEngine, FormulaEngine3Phase)):
             self._steps.append((TokenType.COMPONENT_METRIC, other))
+        elif isinstance(
+            other, (Power, Current, Voltage, Energy, Frequency, Percentage)
+        ):
```
I think using the base class is cleaner and probably requires less maintenance:

```suggestion
        elif isinstance(other, Quantity):
```

shsms commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -879,6 +889,11 @@ def build(
                 assert isinstance(value, str)
                 builder.push_oper(value)
+            elif typ == TokenType.CONSTANT:
+                assert isinstance(
+                    value, (Power, Current, Voltage, Energy, Frequency, Percentage)
+                )
```
same

matthias-wende-frequenz committed:
Add test for composing formulas with constants

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

shsms commented on tests/timeseries/_formula_engine/test_formula_composition.py:
```diff
@@ -170,6 +170,27 @@ async def test_formula_composition_missing_bat(self, mocker: MockerFixture) -> N
+        scaled_grid_power = await engine.new_receiver().receive()
+        assert scaled_grid_power.value is not None
+        assert math.isclose(
+            scaled_grid_power.value._base_value,  # pylint: disable=protected-access
```
you could say `as_watts()` here.

shsms commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -28,6 +28,7 @@
 from frequenz.channels import Broadcast, Receiver
 
 from ..._internal._asyncio import cancel_and_await
+from ...timeseries import Quantity
 from .. import Sample, Sample3Phase
 from .._quantities import QuantityT
```
also I guess Quantity and QuantityT can be imported from the same place.

shsms commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -243,6 +244,8 @@ def __add__(
                 into a formula engine.
         """
+        # if isinstance(other, QuantityT):
+        #     return self._higher_order_builder(self, self._create_method) + other
```
I guess the comment can go away?

matthias-wende-frequenz commented:
updated

shsms reviewed (commented):
Looks good to me,  just a couple of more small comments.

matthias-wende-frequenz committed:
Update release notes

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

shsms commented on tests/timeseries/_formula_engine/test_formula_composition.py:
```diff
@@ -170,6 +170,27 @@ async def test_formula_composition_missing_bat(self, mocker: MockerFixture) -> N
+        )
+
+        await mockgrid.cleanup()
+        await engine._stop()  # pylint: disable=protected-access
```
One last thing,  could you add `await logical_meter.stop()` as well,  so that the `grid_power` engine gets cleaned up,  else there will be weird warning messages in test logs, that make it harder to debug actual issues.

matthias-wende-frequenz replied:
Thanks! good catch. I wanted to do it and than missed it somehow.

shsms commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -771,6 +775,19 @@ def _push(
+                            f"A Quantity must be provided for addition or subtraction to {other}"
+                        )
+                case "*" | "/":
+                    if not isinstance(other, float):
+                        raise RuntimeError(
+                            f"A float must be provided for scalar multiplication to {other}"
+                        )
```
We need an extra check to deal with python weirdness (aka to be idiomatic)

```suggestion
                    if not isinstance(other, float) and not isinstance(other, int):
                        raise RuntimeError(
                            f"A float must be provided for scalar multiplication to {other}"
                        )

matthias-wende-frequenz replied:
fixed

matthias-wende-frequenz committed:
Move formula evaluator to separate file

This patch is done in order to decrease the size of the
`_formula_engine.py` file.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Add failing tests for formula composition

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>
